sys/ztimer/init.c

```diff
@@ -0,0 +1,158 @@
+ *   - if CONFIG_ZTIMER_SEC_TYPE is given
+ *      - set up accordingly
+ *   - if ztimer_msec is using rtt or rtc, use that (convert & extend to seconds)
+ *   - if MODULE_RTT is available, use that
+ *   - if MODULE_RTC is available, use that
+ *   - use ZTIMER_MSEC
+ *
+ * Nested ifdefs are used here as this is all compile-time logic.
+ * This would benefit a lot from code generation...
+ */
+
+#ifdef MODULE_ZTIMER_USEC
+#  ifndef CONFIG_ZTIMER_USEC_TYPE
+#    define CONFIG_ZTIMER_USEC_TYPE ZTIMER_TYPE_PERIPH
+#  endif
+#  if CONFIG_ZTIMER_USEC_TYPE == ZTIMER_TYPE_PERIPH
+#    ifndef CONFIG_ZTIMER_USEC_DEV
+#      define CONFIG_ZTIMER_USEC_DEV    (0)
+#    endif
+#    ifndef CONFIG_ZTIMER_USEC_FREQ
+#      ifdef XTIMER_HZ
+#        define CONFIG_ZTIMER_USEC_FREQ XTIMER_HZ
+#      else
+#        define CONFIG_ZTIMER_USEC_FREQ   (1000000LU)
+#      endif
+#    endif
+#    ifndef CONFIG_ZTIMER_USEC_WIDTH
+#      ifdef XTIMER_WIDTH
+#        define CONFIG_ZTIMER_USEC_WIDTH XTIMER_WIDTH
+#      else
+#        define CONFIG_ZTIMER_USEC_WIDTH (32)
+#      endif
+#    endif
+#    ifndef CONFIG_ZTIMER_USEC_CHAN     /* currently unused! */
+#      define CONFIG_ZTIMER_USEC_CHAN   (0)
+#    endif
+
+#    if CONFIG_ZTIMER_USEC_FREQ == 1000000
+#      define ZTIMER_USEC_DIV           0
+#      define ZTIMER_USEC_MUL           0
+#      define ZTIMER_USEC_CONVERT_BITS  0
+#    elif CONFIG_ZTIMER_USEC_FREQ == 250000
+#      define ZTIMER_USEC_DIV           4
+#      define ZTIMER_USEC_MUL           0
+#      define ZTIMER_USEC_CONVERT_BITS  0
+#    elif CONFIG_ZTIMER_USEC_FREQ == 125000
+#      define ZTIMER_USEC_DIV           16
+#      define ZTIMER_USEC_MUL           0
+#      define ZTIMER_USEC_CONVERT_BITS  0
+#    elif CONFIG_ZTIMER_USEC_FREQ == 62500
+#      define ZTIMER_USEC_DIV           16
+#      define ZTIMER_USEC_MUL           0
+#      define ZTIMER_USEC_CONVERT_BITS  0
+#    elif CONFIG_ZTIMER_USEC_FREQ == 32768
+#      define ZTIMER_USEC_DIV           15625
+#      define ZTIMER_USEC_MUL           512
+       define ZTIMER_USEC_CONVERT_BITS  9
+#    else
+#      error unhandled CONFIG_ZTIMER_USEC_FREQ!
+#    endif
+
+     static ztimer_periph_t _ztimer_usec_periph;
+#    define ZTIMER_USEC_INIT_PERIPH() \
+            ztimer_periph_init(&_ztimer_usec_periph, CONFIG_ZTIMER_USEC_DEV, \
+                             CONFIG_ZTIMER_USEC_FREQ)
+#    define _ZTIMER_USEC_DEV _ztimer_usec_periph
+#  else
+#    error unknown CONFIG_ZTIMER_USEC_TYPE!
+#  endif /* CONFIG_ZTIMER_USEC_TYPE == ZTIMER_TYPE_PERIPH */
+
+# if (CONFIG_ZTIMER_USEC_WIDTH == 32) && (ZTIMER_USEC_CONVERT_BITS == 0)
+    ztimer_dev_t *const ZTIMER_USEC = (ztimer_dev_t *) &_ZTIMER_USEC_DEV;
+# else
+#   if (ZTIMER_USEC_DIV != 0) || (ZTIMER_USEC_MUL != 0)
+      static ztimer_convert_t _ztimer_usec_convert;
+#     define ZTIMER_USEC_INIT_CONVERT() \
+          ztimer_convert_init(&_ztimer_usec_convert, \
+                              (ztimer_dev_t *)&_ZTIMER_USEC_DEV, \
```

Comment:
Shouldn't `&_ZTIMER_USEC_DEV->dev` work regardless of what kind of backend it is, as the dev has to be the first member in any case?